src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -2676,7 +2676,7 @@ private BoundStatement BindReturn(ReturnStatementSyntax syntax, DiagnosticBag di
                 }
             }
 
-            return new BoundReturnStatement(syntax, refKind, arg);
+            return new BoundReturnStatement(syntax, sigRefKind, arg);
```

Comment:
>sigRefKind [](start = 52, length = 10)

Same comment here

Reply:
This is the main issue we need to discuss before I can proceed. There are two reasons I made this change:

1. This code now aligns with the implicit `BoundReturnStatement` constructed for an expression-bodied member
    https://github.com/dotnet/roslyn/blob/7a6034b63cdd5a6712f50f3a284ea5c90443ca08/src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs#L3038
2. `refKind` is never `RefReadOnly`, but downstream consumers are expecting the return statement to reflect the signature ref kind, at least in non-error scenarios

Note that the following error case is likely relevant to code generation scenarios in the IDE:

```csharp
ref readonly int GetSomething() {
  return ref MethodThatDoesNotExist();
}
```

A refactoring to create `MethodThatDoesNotExist` should return `ref readonly int` to reflect the containing signature of the return statement, and not `ref int` that would be the type for the ref-expression.

Reply:
> 1. This code now aligns with the implicit BoundReturnStatement constructed for an expression-bodied member

That was likely unintentional. BTW, it would be good to see if any existing test would fail if the implementation for expression-bodied members would change to align with old behavior for return statement.

> 2. refKind is never RefReadOnly, but downstream consumers are expecting the return statement to reflect the signature ref kind, at least in non-error scenarios

I am fine with that for non error scenarios. However in a case of any mismatch, I think bound node should definitely reflect what is in code. And I also think that IOperation should also reflect what is in code for these cases.

---
In reply to: [292254151](https://github.com/dotnet/roslyn/pull/36271#discussion_r292254151) [](ancestors = 292254151)

Reply:
So if `refKind==RefKind.Ref` and `sigRefKind==RefKind.RefReadOnly`, then we use the latter. And in all other cases we use the former?

Reply:
> So if refKind==RefKind.Ref and sigRefKind==RefKind.RefReadOnly, then we use the latter. And in all other cases we use the former?

Yes.

---
In reply to: [292594829](https://github.com/dotnet/roslyn/pull/36271#discussion_r292594829) [](ancestors = 292594829)

Reply:
It would be good to reflect that in BoundNodes.xml

---
In reply to: [292596858](https://github.com/dotnet/roslyn/pull/36271#discussion_r292596858) [](ancestors = 292596858,292594829)

Reply:
👍 I'll implement *just this piece* in the next step and submit it for incremental review (and CI testing) so we can look at the other feedback in the context of this change.

Reply:
➡️ This is now resolved.